### PR TITLE
feat(data-pipeline): add audio_mp3 preview-column CLI for Lance datasets

### PR DESCRIPTION
**Pull request by ktinubu**
## What

Adds `synth-setter-add-mp3-audio`, a standalone CLI that backfills a per-row
`audio_mp3` **preview** column onto an existing Lance dataset written by
`synth-setter-generate-dataset` or `synth-setter-finalize-dataset`.

- Each row's raw `audio` tensor (float16 `(channels, time)`) is encoded to a CBR
  MP3 with **pedalboard** (already a project dependency — no new deps) and stored
  as a `large_binary` column tagged `mime_type: audio/mpeg`, so open-source Lance
  viewers (e.g. SmooSense) auto-play a per-row preview.
- The column is added via Lance's `batch_udf`-driven `add_columns`, which backfills
  in place in a single transaction without rewriting the existing columns. The
  sample rate is read from the shard's embedded `ShardMetadata`.
- `r2://` URIs are rewritten to `s3://` with env-derived credentials (mirroring
  `finalize_dataset`'s lance path); local paths and `s3://` URIs are used as-is.
- The MP3 is a lossy convenience artifact for auditioning patches — **never a
  training input**; the `audio` tensor stays the source of truth.

Also adds `notebooks/add_mp3_audio.ipynb`: a smoke run that builds a tiny Lance
dataset, runs the add step, and renders a `params + mp3` dataframe.

## Why

A post-hoc batch tool for attaching the MP3 preview column to any already-written
Lance dataset (per-shard or finalized split). This is a different mechanism than
the inline-at-generation/finalize encoding both linked issues propose, so it
partially addresses them and they stay open for the inline work.

It sits alongside the other post-write Lance augmenters (`synth-setter-add-embeddings`)
and follows the same §5 stage contract — adds a column to existing shards without
modifying earlier stages.

Refs #1697
Refs #1682

## Test plan

- 26 unit/integration tests in `tests/pipeline/data/test_add_mp3_audio.py`: MP3
  frame-sync, bitrate-size monotonicity, float16/32/64 dtypes, sample rates
  8k/16k/44.1k, mono + silent + malformed-shape inputs, `mime_type` tag,
  source-column preservation, missing-audio / missing-metadata / duplicate-column
  errors, R2 credentialing for both `r2://` and bare `s3://` URIs (treated as R2,
  matching `add_embeddings`) plus the local-path no-credentials path, and the CLI
  entrypoint (success + non-zero exit).
- Pre-PR gate: `/repo-review-full-no-comments` — 0 BLOCK across 8 skills.

## Follow-ups (advisory, not blocking)

- CLI-level assertion that `--bitrate-kbps` changes encoded size end-to-end
  (currently proven at the unit level).
- Optionally expose `batch_udf(checkpoint_file=...)` for partial-progress
  resumption on very large datasets.

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

Verified on branch `feat/data-pipeline-mp3-audio-column` @ `8db2290f` (= origin HEAD) using the worktree's own `.venv`. Every check invokes the real CLI / system; none parse config.

- [x] **The registered `synth-setter-add-mp3-audio` CLI backfills an `audio_mp3` `large_binary` column (tagged `mime_type: audio/mpeg`) onto a real Lance dataset, leaves the source columns intact, and the bytes decode back to MP3 audio at the dataset's sample rate** *(Level 1 — full end-to-end run of the actual entry point on a dataset built with the repo's own writer helpers; promise is end-to-end)*

  ```bash
  # build a real dataset (sine audio @ 22050 Hz stereo, 4 rows) via lance_shard helpers, then:
  $ uv run synth-setter-add-mp3-audio "$WORK/shard-000000.lance"
  added 'audio_mp3' column to /tmp/.../shard-000000.lance
  # inspect the committed result:
  column type : large_binary
  field meta  : {b'mime_type': b'audio/mpeg'}
  source cols : ['audio', 'mel_spec', 'param_array']
  row count   : 4
  first cell  : 4596 bytes; mp3 sync: 0xff 0xe0
  decoded     : sr= 22050 channels= 2
  df columns  : ['param_array', 'audio_mp3']
  ```

- [x] **Re-running on a dataset that already has the column fails fast with a clear error and non-zero exit** *(Level 1, adversarial — exercises the duplicate-column guard via the CLI)*

  ```bash
  $ uv run synth-setter-add-mp3-audio "$WORK/shard-000000.lance"; echo "exit=$?"
  Error: dataset at /tmp/.../shard-000000.lance already has an 'audio_mp3' column
  exit=1
  ```

- [x] **A dataset with no `audio` column fails fast with a clear error and non-zero exit** *(Level 1, adversarial — exercises the missing-column guard via the CLI)*

  ```bash
  $ uv run synth-setter-add-mp3-audio "$NOAUD/x.lance"; echo "exit=$?"
  Error: dataset at /tmp/.../x.lance has no 'audio' column to encode
  exit=1
  ```

- [x] **The smoke notebook runs top-to-bottom and yields the params + MP3 dataframe** *(Level 1 — executes every notebook code cell in order)*

  ```bash
  $ uv run python -c "exec all code cells of notebooks/add_mp3_audio.ipynb"
  wrote /tmp/mp3-audio-smoke-.../shard-000000.lance
  sample_rate: 16000 channels: 2 frames: 5184
  notebook dataframe columns: ['param_array', 'audio_mp3', 'mp3_bytes']
  notebook rows: 4
  NOTEBOOK RAN TOP-TO-BOTTOM OK
  ```

- [x] **The `synth-setter-add-mp3-audio` entry point is registered and callable** *(Level 2 — invokes the console script; promise is `[project.scripts]` registration)*

  ```bash
  $ uv run synth-setter-add-mp3-audio --help
  Usage: synth-setter-add-mp3-audio [OPTIONS] URI
    Add an ``audio_mp3`` preview column to the Lance dataset at ``URI``.
    --bitrate-kbps INTEGER  Constant MP3 bitrate in kbps applied to every row.
  ```

Unit-level edge cases (silent/mono/malformed audio, float16/32/64, sample rates 8k/16k/44.1k, bitrate-size monotonicity, mime tag, missing-metadata) are covered by the 22 tests in `tests/pipeline/data/test_add_mp3_audio.py`; CI runs the suite.

File: src/synth_setter/pipeline/data/add_mp3_audio.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    is_r2 = r2_io.is_r2_uri(uri)
+    resolved_uri = r2_io.to_s3_uri(uri) if is_r2 else uri
+    storage_options = r2_io.r2_storage_options() if is_r2 else None
+    try:
+        add_mp3_audio_column(
+            resolved_uri, bitrate_kbps=bitrate_kbps, storage_options=storage_options
+        )
+    except (ValueError, OSError) as exc:
+        raise click.ClickException(str(exc)) from exc
+    click.echo(f"added {AUDIO_MP3_FIELD!r} column to {uri}")
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py]: Fixed in 693db186. Moved the R2 URI resolution and `r2_storage_options()` call inside the `try` block and added `RuntimeError` to the handled exceptions, so an `r2://` URI with missing/blank credentials now surfaces a clean `ClickException` instead of a traceback.

File: src/synth_setter/pipeline/data/add_mp3_audio.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        return pa.record_batch([column], names=[AUDIO_MP3_FIELD])
+
+    dataset.add_columns(_to_mp3, read_columns=[AUDIO_FIELD])
+
+
+@click.command()
+@click.argument("uri", type=str)
+@click.option(
+    "--bitrate-kbps",
+    type=int,
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py]: Fixed in 693db186. `--bitrate-kbps` now uses `click.IntRange(min=8, max=320)` (the valid MP3 CBR range), so 0/negative/out-of-range values are rejected at the Click layer. Added `test_main_rejects_out_of_range_bitrate` to pin this.

File: tests/pipeline/data/test_add_mp3_audio.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+def _write_smoke_dataset(path: Path, *, sample_rate: int = _SAMPLE_RATE) -> None:
+    """Write a smoke Lance dataset: real sine ``audio`` plus zeroed mel/param tensors.
+
+    :param path: Destination ``.lance`` directory.
+    :param sample_rate: Sample rate embedded in the shard metadata.
+    """
+    metadata = ShardMetadata(
+        velocity=100,
+        signal_duration_seconds=_DURATION_SECONDS,
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/pipeline/data/test_add_mp3_audio.py]: Fixed in 693db186. `_write_smoke_dataset` now derives `signal_duration_seconds` from `_TIME_SAMPLES / sample_rate`, so the embedded `ShardMetadata` stays consistent with the written tensor shape at any sample rate.

File: src/synth_setter/pipeline/data/add_mp3_audio.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    buffer = io.BytesIO()
+    with AudioFile(
+        buffer,
+        "w",
+        samplerate=sample_rate,
+        num_channels=audio.shape[0],
+        format="mp3",
+        quality=str(bitrate_kbps),
+    ) as out:
+        out.write(audio.astype(np.float32))
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py]: Fixed in 245df2a9. `encode_audio_to_mp3` now uses `np.ascontiguousarray(audio, dtype=np.float32)`, which is a no-op when the input is already float32 and C-contiguous, avoiding the redundant copy while still guaranteeing the encoder gets a contiguous float32 buffer.

File: src/synth_setter/pipeline/data/add_mp3_audio.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    """
+    # Lance opens R2 over its S3-compatible API: rewrite the scheme and pass the
+    # env-derived credentials, mirroring finalize_dataset's lance path. R2
+    # credential resolution raises RuntimeError on missing/blank env, so it stays
+    # inside the try to surface as a clean ClickException.
+    try:
+        is_r2 = r2_io.is_r2_uri(uri)
+        resolved_uri = r2_io.to_s3_uri(uri) if is_r2 else uri
+        storage_options = r2_io.r2_storage_options() if is_r2 else None
+        add_mp3_audio_column(
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py]: Fixed in 245df2a9. Added `test_main_rewrites_r2_uri_and_forwards_storage_options`, which monkeypatches `r2_io.is_r2_uri`/`to_s3_uri`/`r2_storage_options`, spies on `add_mp3_audio_column`, invokes `main` with an `r2://` URI, and asserts the rewritten `s3://` URI and storage options are forwarded.

File: tests/pipeline/data/test_add_mp3_audio.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    """Build ``(_ROWS, _CHANNELS, T)`` float16 sine audio, one pitch per row.
+
+    :param sample_rate: Rate used to space the time axis so the tone is real.
+    :returns: Float16 audio with a distinct frequency per row.
+    """
+    t = np.arange(_TIME_SAMPLES) / sample_rate
+    rows = [np.sin(2 * np.pi * freq * t) for freq in (220.0, 440.0, 660.0)]
+    mono = np.stack(rows)[:, None, :]
+    # broadcast_to returns a read-only view; copy before mutating a returned row.
+    return np.broadcast_to(mono, (_ROWS, _CHANNELS, _TIME_SAMPLES)).astype(np.float16)
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/pipeline/data/test_add_mp3_audio.py]: Fixed in 245df2a9. Reworded the comment to note that the trailing `.astype(np.float16)` materializes a fresh writable array, so the returned value is safe to mutate (no defensive copy needed by callers).

**Review by khaledtin — COMMENTED** [review]
Re-post of repo-review-full WARN findings from PR #1718 as individual unresolved inline threads. The original orchestrator returned the lance-review report inline instead of routing through `post_review.py`; this re-post anchors each WARN to its file:line so the threads stay visible and require explicit resolution. Source skill: lance-review (pylance 7.0.0). No BLOCK findings.

File: src/synth_setter/pipeline/data/add_mp3_audio.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    )
+
+    # output_schema makes Lance skip the first-batch inference probe (which would
+    # encode that batch twice) and carries the mime metadata onto the new column.
+    @lance.batch_udf(output_schema=mp3_schema)
+    def _to_mp3(batch: pa.RecordBatch) -> pa.RecordBatch:
+        column = _encode_audio_column(batch, sample_rate, bitrate_kbps)
+        return pa.record_batch([column], names=[AUDIO_MP3_FIELD])
+
+    dataset.add_columns(_to_mp3, read_columns=[AUDIO_FIELD])
```
Comment (khaledtin) [inline src/synth_setter/pipeline/data/add_mp3_audio.py:126]: **[lance:warn]** `add_columns` is invoked without a `checkpoint_file`, so a partial failure (e.g. cloud I/O hiccup mid-encode of a multi-GB shard) loses every previously-encoded batch and re-encoding restarts from scratch. The PR's own docstring on lines 97-100 names exactly the scenario the documented checkpoint primitive exists to mitigate. Docs: *"If the function is expensive to compute and can fail, it is recommended to set a checkpoint file in the UDF. This checkpoint file saves the state of the UDF after each invocation, so that if the UDF fails, it can be restarted from the last checkpoint."* — https://lance.org/guide/data_evolution/ (Adding columns / batch_udf section). Threading a `--checkpoint-file` Click option through `add_mp3_audio_column` to the `@lance.batch_udf(...)` decorator lets long R2-backed runs resume; the in-process file can be discarded on success.

File: src/synth_setter/pipeline/data/add_mp3_audio.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    :raises ValueError: ``uri`` lacks an ``audio`` column, already has an
+        ``audio_mp3`` column, or carries no readable shard metadata.
+    """
+    dataset = lance.dataset(str(uri), storage_options=storage_options)
+    if AUDIO_FIELD not in dataset.schema.names:
+        raise ValueError(f"dataset at {uri} has no {AUDIO_FIELD!r} column to encode")
+    if AUDIO_MP3_FIELD in dataset.schema.names:
+        raise ValueError(f"dataset at {uri} already has an {AUDIO_MP3_FIELD!r} column")
+    sample_rate = read_shard_metadata(dataset.schema).sample_rate
+    mp3_schema = pa.schema(
```
Comment (khaledtin) [inline src/synth_setter/pipeline/data/add_mp3_audio.py]: **[lance:warn]** The preview column is added as a plain `pa.large_binary()` with a `mime_type` field-metadata hint, but the repo writes datasets at `data_storage_version="2.2"` (`pipeline/data/lance_shard.py:21`), where Lance ships a first-class blob path (`blob_field` / `blob_array` / `take_blobs`) that supports lazy per-row fetch — the exact access pattern an MP3-preview column wants (read the row's MP3 without materializing the full column on every scan). The page explicitly steers v2.2+ datasets toward the modern API: *"For new datasets, prefer blob v2."* — https://lance.org/guide/blob/. `pa.large_binary()` is still correct and writeable; this is a forward-looking simplification, not a correctness gap. The `mime_type: audio/mpeg` metadata convention is not Lance-documented (the blob guide doesn't define a mime_type field-metadata contract), so its viewer behavior is a repo convention, not a Lance guarantee — worth a one-line comment noting that on line 29-31.

File: src/synth_setter/pipeline/data/add_mp3_audio.py (hunk 7 above)
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py:126]: Deferred — tracked in #1729. `add_columns` runs as a single Lance transaction, so an interrupted run leaves the dataset on its prior version and re-running is safe; correctness isn't at risk. A `checkpoint_file` only saves *recompute* cost on a failed retry, which matters once shard encodes get expensive enough that losing partial progress hurts (multi-GB shards / flaky cloud I/O). Unnecessary for the MVP — we'll add it if/when we hit real mid-encode failures.

File: src/synth_setter/pipeline/data/add_mp3_audio.py (hunk 8 above)
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py]: Deferred — tracked in #1729. One correction: the repo pins `LANCE_DATA_STORAGE_VERSION = "2.1"` (`src/synth_setter/pipeline/data/lance_shard.py:22`), not 2.2, and Lance's first-class blob path (`blob_field`/`blob_array`/`take_blobs`) requires `data_storage_version >= 2.2` — so it isn't available today. When the storage floor is raised to >= 2.2 we'll migrate the `audio_mp3` column to `blob_field` for lazy per-row fetch.

Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py]: Done in 04ff84d9 — switched `audio_mp3` to Lance blob v2: schema via `lance.blob_field(AUDIO_MP3_FIELD).with_metadata(_MP3_FIELD_METADATA)`, the `batch_udf` returns `lance.blob_array(...)`, and reads go through `dataset.take_blobs(blob_column=..., indices=[...]).readall()` (plain `to_table` yields blob descriptors, not bytes). The branch is now rebased onto `main`, so the dataset storage floor is 2.2 (#1715) which blob v2 requires. Kept the `mime_type: audio/mpeg` field metadata and added a comment noting it is a repo viewer convention, not a Lance-defined contract.

File: src/synth_setter/pipeline/data/add_mp3_audio.py (hunk 7 above)
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py:126]: Deferred to #1731. `add_columns` is atomic — an interrupted run leaves the dataset on its prior version (no partial commit), so re-running is always safe; the cost of a mid-run failure is recompute, not data loss. Threading `--checkpoint-file` for resumable multi-GB R2 encodes is a real improvement but orthogonal to adding the column, so I tracked it as a scoped follow-up (#1731) rather than expanding this PR.

File: src/synth_setter/pipeline/data/add_mp3_audio.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    """Encode a batch's ``audio`` tensor column into a Lance blob array of MP3 bytes.
+
+    :param batch: Record batch projecting the ``audio`` fixed-shape tensor column.
+    :param sample_rate: Playback rate in Hz every row is encoded at.
+    :param bitrate_kbps: Constant MP3 bitrate in kbps for every row.
+    :returns: One Lance blob array of per-row MP3 bytes, in batch row order.
+    :raises ValueError: The ``audio`` column is not a fixed-shape tensor column,
+        or a row fails to encode (the message names the offending row index).
+    """
+    column = batch.column(AUDIO_FIELD)
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py:75]: Not a bug — `pa.RecordBatch.column()` accepts **either** an integer index **or** a string field name. The bound signature is `_Tabular.column(self, i)`, and `i` is resolved by name when a string is passed (`rb.column("audio")` returns the column on PyArrow '>=14). This exact call (`batch.column(AUDIO_FIELD)`) is exercised end-to-end by `test_add_mp3_audio_column_adds_blob_column_for_every_row`, which runs the full `add_columns`/`batch_udf` path and decodes the result — so a `TypeError` here would fail CI. `lance_shard.iter_lance_column_rows` uses `batch.column(0)` simply because it projects a single column, not because string lookup is unsupported. Leaving as-is.

File: src/synth_setter/pipeline/data/add_mp3_audio.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    # env-derived credentials, mirroring finalize_dataset's lance path. R2
+    # credential resolution raises RuntimeError on missing/blank env, so it stays
+    # inside the try to surface as a clean ClickException.
+    try:
+        is_r2 = r2_io.is_r2_uri(uri)
+        resolved_uri = r2_io.to_s3_uri(uri) if is_r2 else uri
+        storage_options = r2_io.r2_storage_options() if is_r2 else None
+        add_mp3_audio_column(
+            resolved_uri, bitrate_kbps=bitrate_kbps, storage_options=storage_options
+        )
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply src/synth_setter/pipeline/data/add_mp3_audio.py:166]: Fixed in f697630d. `main()` now mirrors `add_embeddings._open_lance_dataset`: after rewriting any `r2://` to `s3://`, it treats **any** `s3://` URI as the project's R2 endpoint — calling `ensure_r2_env_loaded()` and attaching `r2_storage_options()` — so `synth-setter-add-mp3-audio s3://…` is credentialed from `RCLONE_CONFIG_R2_*` rather than relying on ambient AWS env. Added `test_main_credentials_bare_s3_uri_as_r2` (and `test_main_local_path_passes_no_storage_options` to pin the local path). Matching the sibling's documented contract, generic non-R2 S3 buckets remain unsupported.

File: tests/pipeline/data/test_add_mp3_audio.py [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+    ]
+
+
+def test_encode_audio_to_mp3_starts_with_frame_sync() -> None:
+    """A valid MP3 stream opens with the 11-bit frame-sync prefix (0xFF 0xEy/0xFy)."""
+    payload = encode_audio_to_mp3(_sine_rows()[0], _SAMPLE_RATE, 128)
+
+    assert len(payload) > 0
+    assert payload[0] == 0xFF
+    assert payload[1] & 0xE0 == 0xE0
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/pipeline/data/test_add_mp3_audio.py]: Fixed in bb5c5135. The test now scans the first 1 KB for the frame-sync word (`0xFF`, `0xE0` masked) instead of asserting it at offset 0, and is renamed `test_encode_audio_to_mp3_contains_frame_sync` — so a payload that legally leads with an ID3 tag or padding still passes as long as it carries a decodable frame.

File: docs/doc-map.yaml [hunk 12]
```diff
@@ -116,6 +116,8 @@ docs:
         covers: "Lance dataset-shard codec — lance_schema / tensor_array / record_batch_from_arrays / write_lance_dataset / lance_fragment / commit_lance_dataset on the encode side, read_shard_metadata / iter_lance_column_rows on the decode side; shared by the Lance writer, validator, and finalize."
       - pattern: "src/synth_setter/pipeline/data/add_embeddings.py"
         covers: "`synth-setter-add-embeddings` CLI — appends a `clap` (LAION-CLAP) `FixedSizeList<float32,512>` vector column (with an optional IVF_PQ index for `nearest=` search) and an `m2l` (music2latent) fixed-shape-tensor column to a finalized Lance dataset via `lance.batch_udf`, with injected/lazy-loaded encoders and row-count + dim + finiteness validation (post-finalize augmenter)."
+      - pattern: "src/synth_setter/pipeline/data/add_mp3_audio.py"
+        covers: "synth-setter-add-mp3-audio CLI — backfills an audio_mp3 large_binary preview column (CBR MP3 via pedalboard, tagged mime_type: audio/mpeg) onto a Lance dataset via Lance batch_udf add_columns; a lossy auditioning preview, never a training input."
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply docs/doc-map.yaml]: Fixed in bb5c5135. Reworded the doc-map entry to call `audio_mp3` a Lance blob v2 column (`lance.blob_field` / `lance.blob_array`) and backticked the CLI/name/type tokens to match the adjacent `add_embeddings` / `add_music2latent` entries.
